source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,583 @@
+                    for (unsigned int i=0; i<dofs_per_cell; ++i)
+                      {
+                        // Make sure we only assemble for the y-component
+                        // TODO this is only correct for box geometries
+                        if (mesh_deformation_dof_handler.get_fe().system_to_component_index(i).first != dim-1)
+                          continue;
+
+                        // Assemble the RHS
+                        // RHS = M*H_old
+                        cell_vector(i) += displacement *
+                                          fs_fe_face_values.shape_value (i, point) *
+                                          fs_fe_face_values.JxW(point);
+
+
+                        for (unsigned int j=0; j<dofs_per_cell; ++j)
+                          {
+                            // Make sure we only assemble for the y-component
+                            // TODO this is only correct for box geometries
+                            if (mesh_deformation_dof_handler.get_fe().system_to_component_index(j).first != dim-1)
+                              continue;
+                            // Assemble the matrix, for backward first order time discretization:
+                            // Matrix := (M+dt*K) = (M+dt*B^T*kappa*B)
+                            cell_matrix(i,j) +=
+                              (
+                                this->get_timestep() * diffusivity *
+                                projection * fs_fe_face_values.shape_grad(i, point) * projection * fs_fe_face_values.shape_grad(j,point) +
+                                fs_fe_face_values.shape_value (i, point) * fs_fe_face_values.shape_value (j, point)
+                              )
+                              * fs_fe_face_values.JxW(point);
+                          }
+                      }
+
+                  }
+
+                mass_matrix_constraints.distribute_local_to_global (cell_matrix, cell_vector,
+                                                                    cell_dof_indices, mass_matrix, system_rhs, false);
+              }
+
+      system_rhs.compress (VectorOperation::add);
+      mass_matrix.compress(VectorOperation::add);
+
+      // Jacobi seems to be fine here.  Other preconditioners (ILU, IC) run into trouble
+      // because the matrix is mostly empty, since we don't touch internal vertices.
+      LinearAlgebra::PreconditionJacobi preconditioner_mass;
+      LinearAlgebra::PreconditionJacobi::AdditionalData preconditioner_control(1,1e-16,1);
+      preconditioner_mass.initialize(mass_matrix, preconditioner_control);
+
+      this->get_pcout() << "   Solving mesh surface diffusion" << std::endl;
+      SolverControl solver_control(50*system_rhs.size(), this->get_parameters().linear_stokes_solver_tolerance*system_rhs.l2_norm());
+      SolverCG<LinearAlgebra::Vector> cg(solver_control);
+      cg.solve (mass_matrix, solution, system_rhs, preconditioner_mass);
```

Comment:
I think I do the same as in the free surface implementation, which I copy here below for ease:
```
      // Jacobi seems to be fine here.  Other preconditioners (ILU, IC) run into troubles
      // because the matrix is mostly empty, since we don't touch internal vertices.
      LinearAlgebra::PreconditionJacobi preconditioner_mass;
      preconditioner_mass.initialize(mass_matrix);

      SolverControl solver_control(5*rhs.size(), this->get_parameters().linear_stokes_solver_tolerance*rhs.l2_norm());
      SolverCG<LinearAlgebra::Vector> cg(solver_control);
      cg.solve (mass_matrix, dist_solution, rhs, preconditioner_mass);
```

However, this gave me very unstable results, with only certain solver settings leading to solver convergence. This was fixed when I added
```
      LinearAlgebra::PreconditionJacobi::AdditionalData preconditioner_control(1,1e-16,1);
```
which takes care of the zero entries in the matrix. See the discussion above your comments.

Reply:
I remember that last hack we looked at the free surface implementation with Timo as well, and decided it could be better (see issue #3110), but works for now.